src/Servers/HttpSys/src/RequestProcessing/Response.cs

```diff
@@ -400,7 +412,7 @@ internal HttpApiTypes.HTTP_FLAGS ComputeHeaders(long writeCount, bool endOfReque
                 _boundaryType = BoundaryType.ContentLength;
                 // ComputeLeftToWrite checks for HEAD requests when setting _leftToWrite
                 _expectedBodyLength = responseContentLength.Value;
-                if (_expectedBodyLength == writeCount && !isHeadRequest)
```

Comment:
Does this if a Content-Length response header is set, you have to also set a Trailer response header to be allowed send trailers? Can you still send trailers if you don't set the Trailer response header but you write the response in multiple chunks instead of a single write? Or if you add to the trailers dictionary before a single write? 

If so, that behavior is probably going to surprise developers. If you have to set a Trailer response header in some cases, you should have to send it all cases.

Reply:
Discussed offline. You either have to declare trailers or add at least one trailer before the last write. The number of writes doesn't affect the logic (that code is just about kernel mode caching).

 

> If so, that behavior is probably going to surprise developers.

Yeah. Maybe I should make the collection readonly when it's not going to work so you get clearer guidance.

> If you have to set a Trailer response header in some cases, you should have to send it all cases.

No, the more common cases don't need it. I looked through the gRPC usage with John and found that it does not declare Trailers and it would be difficult for them to do so. E.g. the status can be sent as a normal response header if there is no body or as a trailer if there is one. And the message response trailer they don't decide if it's needed until after the end of the body. Note gRPC doesn't use Content-Length so they aren't affected by these more nuanced scenarios. I wouldn't want to make their work harder just to ensure consistency for the less common Content-Length scenario.